clang/lib/AST/ExprConstant.cpp

```diff
@@ -10895,6 +10899,132 @@ bool VectorExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+      return false;
+  }
+
+  return Result.isVector();
+}
+
+static bool handleVectorConversion(EvalInfo &Info, const FPOptions FPO,
+                                   const Expr *E, QualType SourceTy,
+                                   QualType DestTy, APValue const &Original,
+                                   APValue &Result) {
+  if (SourceTy->isIntegerType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = APValue(APFloat(0.0));
+      return HandleIntToFloatCast(Info, E, FPO, SourceTy, Original.getInt(),
+                                  DestTy, Result.getFloat());
+    }
+    if (DestTy->isIntegerType()) {
+      Result = APValue(
+          HandleIntToIntCast(Info, E, DestTy, SourceTy, Original.getInt()));
+      return true;
+    }
+  } else if (SourceTy->isRealFloatingType()) {
+    if (DestTy->isRealFloatingType()) {
+      Result = Original;
+      return HandleFloatToFloatCast(Info, E, SourceTy, DestTy,
+                                    Result.getFloat());
+    }
+    if (DestTy->isIntegerType()) {
+      Result = APValue(APSInt());
+      return HandleFloatToIntCast(Info, E, SourceTy, Original.getFloat(),
+                                  DestTy, Result.getInt());
+    }
+  }
+  return false;
+}
+
+bool VectorExprEvaluator::VisitConvertVectorExpr(const ConvertVectorExpr *E) {
+  APValue Source;
+  QualType SourceVecType = E->getSrcExpr()->getType();
+  if (!EvaluateVectorOrLValue(Source, Info, E->getSrcExpr(), SourceVecType))
+    return false;
+
+  QualType DestTy = E->getType()->castAs<VectorType>()->getElementType();
+  QualType SourceTy = SourceVecType->castAs<VectorType>()->getElementType();
+
+  const FPOptions FPO = E->getFPFeaturesInEffect(Info.Ctx.getLangOpts());
+
+  auto SourceLen = Source.getVectorLength();
+  SmallVector<APValue, 4> ResultElements;
+  ResultElements.reserve(SourceLen);
+  for (unsigned EltNum = 0; EltNum < SourceLen; ++EltNum) {
+    APValue Elt;
+    if (!handleVectorConversion(Info, FPO, E, SourceTy, DestTy,
+                                Source.getVectorElt(EltNum), Elt))
+      return false;
+    ResultElements.push_back(std::move(Elt));
+  }
+
+  return Success(APValue(ResultElements.data(), ResultElements.size()), E);
+}
+
+static bool handleVectorShuffle(EvalInfo &Info, const ShuffleVectorExpr *E,
+                                QualType ElemType, APValue const &VecVal1,
+                                APValue const &VecVal2, unsigned EltNum,
+                                APValue &Result) {
+  unsigned const TotalElementsInAVector = VecVal1.getVectorLength();
+
+  Expr const *IndexExpr = E->getExpr(2 + EltNum);
+  APSInt IndexVal;
+  if (!EvaluateInteger(IndexExpr, IndexVal, Info))
+    return false;
+
+  uint32_t index = IndexVal.getZExtValue();
+  // The spec says that -1 should be treated as undef for optimizations,
+  // but in constexpr we need to choose a value. We'll choose 0.
+  if (index >= TotalElementsInAVector * 2)
+    index = 0;
```

Comment:
I don't think this quite implements `__builtin_shufflevector`; consider:

```c++
typedef char vector4char __attribute__((__vector_size__(4)));
typedef char vector8char __attribute__((__vector_size__(8)));

__builtin_shufflevector((vector4char){}, (vector8char){}, 10);
```

That should produce a single-element `char` vector whose element has value `0`: https://godbolt.org/z/9655x1oc6 (though it looks like there's another issue elsewhere in Sema that's not currently accepting that).

Also, the `>=` suggests to me that it'll also produce values for out-of-bounds indexes. Happily, this already aborts (with a useful diagnostic):

```c++
__builtin_shufflevector((vector8char){}, (vector8char){}, 20);
```

So, for that part it's probably enough here to clarify with an `assert` that `index` is less the sum of the vectors' lengths (or the -1 sentinel).

Reply:
Also, I think just picking a value here isn't quite right. Consider:

```c++
constexpr auto f() {
    return __builtin_shufflevector((vector8char){7}, (vector8char){}, -1);
}
```

This implementation will always produce a value of `7` for `f()[0]` when called in a constant context, right? But it's some flavor of U{nspecified,ndefined}B, so the value produced at runtime is unknown. As I understand it the evaluator is required to bail out on UB, to prevent exactly that kind of mismatch: it's preferable for the interpreter to say "I can't do anything with this" than to do something that would only match some of the time.

So, as I understand it, there's two options here:

1. `return false` with some diagnostic saying "no `-1` indexes in a constant context", or
2. use `APValue::Indeterminate` (or `APValue::None`?) for this slot in the vector (oh, I guess _that's_ why there's an APValue per element).

The latter would allow an expression like:

```c++
static_assert(__builtin_shufflevector((vector8char){7}, (vector8char){}, 0, -1)[0] == 7);
```

but not

```c++
// should fail to produce a constant value
static_assert(__builtin_shufflevector((vector8char){7}, (vector8char){}, 0, -1)[1]); 
```

(interestingly, neither `gcc` nor `clang` works completely as I'd expect here; clang doesn't yet support constant-time `[..]` vector element accesses, and `gcc` actually allows the second construction to produce `0`, which it probably shouldn't).

Reply:
<source>:6:12: error: first two arguments to '__builtin_shufflevector' must have the same type
    6 |     return __builtin_shufflevector((vector4char){}, (vector8char){}, 10)[0];
      |            ^                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

Latest trunk seems to also think that both input vectors should match in size. I'll add an llvm_unreachable for the case where it is out of bounds

Reply:
APValue::Indeterminate seems to be exactly what I was looking for! I'll read up on this, thank you. I assumed that such a thing did not exist, but I'm glad to be proven wrong

Reply:
Operator[] needs lvalue support for vectors, I think I commented with @RKSimon that it would be a really useful thing for validations. For now, I'm static casting to get arround this

Reply:
Bad news... seems like APValue::Indeterminate and None are both intentionally disallowed. I'll go with the second option then, make the user choose what element they want to use as a source in constexpr. They can always `if consteval`

Reply:
re `Operator[]` there's already a PR for doing that in constant contexts, I think: https://github.com/llvm/llvm-project/pull/72607 . Definitely agree it'd be helpful for validations! 

re: `APValue::Indeterminate`, I think you're talking about how top-level indeterminate values get rejected by clang as not being a constant expression, right? Dang, that both makes sense and feels unfortunate: thanks for giving it a try, at least!

Reply:
Great, let's hope operator[] makes it.

The error comes from somewhere in the depths of ExprConst, it intentionally tells you that undefined values are indeed illegal in constexpr code. I agree with you, a shame it makes sense!